common/src/main/java/tachyon/dag/DAG.java

```diff
@@ -0,0 +1,173 @@
+        newNode.addParent(parentNode);
+      }
+    }
+  }
+
+  public void deleteLeaf(T payload) {
+    Preconditions.checkState(contains(payload), "the node does not exist");
+    DAGNode<T> node = mIndex.get(payload);
+    Preconditions.checkState(node.getChildren().isEmpty(), "the node is not a leaf");
+
+    // delete from parent
+    for (DAGNode<T> parent : node.getParents()) {
+      parent.removeChild(node);
+    }
+
+    // remove from index
+    mIndex.remove(payload);
+
+    if (node.getParents().isEmpty()) {
+      mRoots.remove(node);
+    }
+  }
+
+  /**
+   * @return true if there a node in the DAG contains the given value as payload, false otherwise.
+   */
+  public boolean contains(T payload) {
+    return mIndex.containsKey(payload);
+  }
+
+  /**
+   * @return the children's payloads, null if the given payload doesn't exist in the DAG.
+   */
+  public List<T> getChildren(T payload) {
+    if (!mIndex.containsKey(payload)) {
+      return null;
+    }
+    DAGNode<T> node = mIndex.get(payload);
+    List<T> children = Lists.newArrayList();
+    for (DAGNode<T> child : node.getChildren()) {
+      children.add(child.getPayload());
+    }
+    return children;
+  }
+
+  /**
+   * @return the parents' payloads, null if the given payload doesn't exist in the DAG.
+   */
+  public List<T> getParents(T payload) {
+    if (!mIndex.containsKey(payload)) {
+      return null;
+    }
+    DAGNode<T> node = mIndex.get(payload);
+    List<T> parents = Lists.newArrayList();
+    for (DAGNode<T> parent : node.getParents()) {
+      parents.add(parent.getPayload());
+    }
+    return parents;
+  }
+
+  /**
+   * @return true if the payload is in the root of the DAG, false otherwise.
+   */
+  public boolean isRoot(T payload) {
+    if (!contains(payload)) {
+      return false;
+    }
+    return mRoots.contains(mIndex.get(payload));
+  }
+
+  /**
+   * @return all the root payloads.
+   */
+  public List<T> getRoots() {
+    List<T> roots = Lists.newArrayList();
+    for (DAGNode<T> root : mRoots) {
+      roots.add(root.getPayload());
+    }
+    return roots;
+  }
+
+  /**
+   * Sorts a given set of payloads topologically based on the DAG. This method requires all the
+   * payloads to be in the DAG. TODO(yupeng): optimize this implementation
+   *
+   * @param payloads the set of input payloads
+   * @return the payloads after topogological sort
+   */
+  public List<T> sortTopologically(Set<T> payloads) {
+    List<T> result = Lists.newArrayList();
+
+    Set<T> input = Sets.newHashSet(payloads);
+    Deque<DAGNode<T>> toVisit = new ArrayDeque<DAGNode<T>>(mRoots);
+    while (!toVisit.isEmpty()) {
+      DAGNode<T> visit = toVisit.removeFirst();
+      T payload = visit.getPayload();
+      if (input.remove(payload)) {
+        result.add(visit.getPayload());
+      }
+      toVisit.addAll(visit.getChildren());
+    }
+
+    Preconditions.checkState(toVisit.isEmpty(), "Not all the given payloads are in the DAG: ",
```

Comment:
Is this a possible case if everything on our side is correct?
If it is just a sanity check, I prefer to leave it in unit test.